crates/zkevm_circuits/src/bn254/ec_mul/implementation.rs

```diff
@@ -0,0 +1,428 @@
+        },
+        tables::ByteSplitTable,
+        u16::UInt16,
+        u512::UInt512,
+    },
+    pairing::ff::PrimeField,
+};
+
+use super::*;
+
+// -- GLV constants --
+
+// Width 4 windowed multiplication parameters
+const WINDOW_WIDTH: usize = 4;
+const NUM_MULTIPLICATION_STEPS_FOR_WIDTH_4: usize = 33;
+const PRECOMPUTATION_TABLE_SIZE: usize = (1 << WINDOW_WIDTH) - 1;
+
+/// BETA parameter such that phi(x, y) = (beta*x, y)
+/// is a valid endomorphism for the curve. Note
+/// that it is possible to use one since 3 divides prime order - 1.
+/// Detailed explanation can be found in file `endomorphism.sage` in `sage` folder.
+const BETA: &str = "2203960485148121921418603742825762020974279258880205651966";
+
+// Decomposition constants for vectors (a1, b1) and (a2, b2),
+// derived through algorithm 3.74 http://tomlr.free.fr/Math%E9matiques/Math%20Complete/Cryptography/Guide%20to%20Elliptic%20Curve%20Cryptography%20-%20D.%20Hankerson,%20A.%20Menezes,%20S.%20Vanstone.pdf
+// Also see `balanced_representation.sage` file for details
+
+/// `a1` component of a short vector `v1=(a1, b1)`.
+const A1: &str = "0x89d3256894d213e3";
+/// `-b1` component of a short vector `v1=(a1, b1)`.
+/// Since `b1` is negative, we use `-b1` instead of `b1`.
+const B1: &str = "0x6f4d8248eeb859fc8211bbeb7d4f1128";
+/// `a2` component of a short vector `v2=(a2, b2)`.
+const A2: &str = "0x6f4d8248eeb859fd0be4e1541221250b";
+/// `b2` component of a short vector `v2=(a2, b2)`.
+const B2: &str = "0x89d3256894d213e3";
+// Note: `a1` and `a2` are not currently used in 4 bit window-based multiplication method.
+
+/// Precomputed value of `-b1/n << 256`
+const G1: &str = "0x24ccef014a773d2cf7a7bd9d4391eb18d";
+/// Precomputed value of `b2/n << 256`
+const G2: &str = "0x2d91d232ec7e0b3d7";
+
+/// Lambda parameter
+const LAMBDA: &str = "0xb3c4d79d41a917585bfc41088d8daaa78b17ea66b99c90dd";
+
+// -- WNAF parameters --
+
+/// The scalar multiplication window size.
+const GLV_WINDOW_SIZE: usize = 2;
+/// The GLV table length.
+const L: usize = 1 << (GLV_WINDOW_SIZE - 1);
+
+/// Converts the `UInt256<F>` element to a non-native field element over `u16`.
+pub fn convert_uint256_to_field_element<F, CS, P, const N: usize>(
+    cs: &mut CS,
+    value: &UInt256<F>,
+    params: &Arc<NonNativeFieldOverU16Params<P, N>>,
+) -> NonNativeFieldOverU16<F, P, N>
+where
+    F: SmallField,
+    CS: ConstraintSystem<F>,
+    P: PrimeField,
+{
+    // We still have to decompose it into u16 words
+    let zero_var = Num::allocated_constant(cs, F::ZERO).get_variable();
+    let mut limbs = [zero_var; N];
+    assert!(N >= 16);
+
+    for (dst, src) in limbs.array_chunks_mut::<2>().zip(value.inner.iter()) {
+        let [b0, b1, b2, b3] = src.to_le_bytes(cs);
+        let low = UInt16::from_le_bytes(cs, [b0, b1]);
+        let high = UInt16::from_le_bytes(cs, [b2, b3]);
+
+        *dst = [low.get_variable(), high.get_variable()];
+    }
+
+    let mut max_value = U1024::from_word(1u64);
```

Comment:
You can reuse params that you created at the beginning of precompile_inner_

Reply:
If you are about `max_value`, it is unconstrained, or do you mean something else? Basically, we have taken this [function from ecrecover](https://github.com/distributed-lab/zksync-protocol/blob/dl-precompiles/crates/zkevm_circuits/src/ecrecover/baseline.rs#L83)